File: src/lib/crew.ts
/** Crew roster helpers (Phase 19a) — mirrors ATS DEFAULT_CREW_LIST + aliases. */

export type CrewMember = {
  name: string;
  tandemmaster: boolean;
  videospringer: boolean;
  aliases: string[];
};

export const DEFAULT_CREW_LIST: CrewMember[] = [
  {name: "Alberto", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Ana", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Andy", tandemmaster: true, videospringer: true, aliases: ["Andreas"]},
  {name: "Chris", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Cornelius", tandemmaster: true, videospringer: false, aliases: ["Corni"]},
  {name: "Futti", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Harry", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Henrik", tandemmaster: true, videospringer: true, aliases: ["Henni"]},
  {name: "Jan", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Jojo", tandemmaster: false, videospringer: true, aliases: []},
  {name: "Kai", tandemmaster: false, videospringer: true, aliases: []},
  {name: "Käthe", tandemmaster: false, videospringer: true, aliases: []},
  {name: "Mathi", tandemmaster: true, videospringer: true, aliases: ["Mathias"]},
  {name: "Max", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Mayo", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Pascal", tandemmaster: true, videospringer: false, aliases: ["Passy"]},
  {name: "Ralph", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Rene", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Robert", tandemmaster: false, videospringer: true, aliases: []},
  {name: "Robin", tandemmaster: false, videospringer: true, aliases: []},
  {name: "Sabrina", tandemmaster: false, videospringer: true, aliases: []},
  {name: "Sahira", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Samuel", tandemmaster: true, videospringer: true, aliases: ["Samu"]},
  {name: "Stefan", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Steve", tandemmaster: true, videospringer: false, aliases: []},
  {name: "Tim", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Tom", tandemmaster: true, videospringer: true, aliases: []},
  {name: "Torsten", tandemmaster: true, videospringer: true, aliases: []},
].sort((a, b) => a.name.localeCompare(b.name, "de"));

function normalizeMember(raw: unknown): CrewMember | null {
  if (!raw || typeof raw !== "object") return null;
  const o = raw as Record<string, unknown>;
  const name = typeof o.name === "string" ? o.name.trim() : "";
  if (!name) return null;
  const aliases = Array.isArray(o.aliases)
    ? o.aliases
        .filter((a): a is string => typeof a === "string")
        .map((a) => a.trim())
        .filter(Boolean)
    : [];
  return {
    name,
    tandemmaster: Boolean(o.tandemmaster),
    videospringer: Boolean(o.videospringer),
    aliases,
  };
}

/** Parse `crew_list` setting JSON; empty/invalid → defaults. */
export function parseCrewList(raw: string | null | undefined): CrewMember[] {
  const trimmed = (raw ?? "").trim();
  if (!trimmed) return DEFAULT_CREW_LIST.map(cloneCrewMember);
  try {
    const parsed = JSON.parse(trimmed) as unknown;
    if (!Array.isArray(parsed) || parsed.length === 0) {
      return DEFAULT_CREW_LIST.map(cloneCrewMember);
    }
    const list = parsed
      .map(normalizeMember)
      .filter((m): m is CrewMember => m !== null);
    return list.length > 0 ? list : DEFAULT_CREW_LIST.map(cloneCrewMember);
  } catch {
    return DEFAULT_CREW_LIST.map(cloneCrewMember);
  }
}

export function serializeCrewList(list: CrewMember[]): string {
  return JSON.stringify(list);
}

export function cloneCrewMember(m: CrewMember): CrewMember {
  return {
    name: m.name,
    tandemmaster: m.tandemmaster,
    videospringer: m.videospringer,
    aliases: [...m.aliases],
  };
}

export function emptyCrewDraft(): CrewMember {
  return {name: "", tandemmaster: true, videospringer: false, aliases: []};
}

export function crewNamesEqual(
  a: string | null | undefined,
  b: string | null | undefined,
): boolean {
  const left = (a ?? "").trim().toLowerCase();
  const right = (b ?? "").trim().toLowerCase();
  return Boolean(left) && left === right;
}

export function crewNamesForRole(
  list: CrewMember[],
  role: "tandemmaster" | "videospringer",
): string[] {
  return list
    .filter((c) => (role === "tandemmaster" ? c.tandemmaster : c.videospringer))
    .map((c) => c.name)
    .sort((a, b) => a.localeCompare(b, "de"));
}

/** Ensure the current value appears in options even if not in the filtered role list. */
export function crewSelectOptions(
  list: CrewMember[],
  role: "tandemmaster" | "videospringer",
  current: string | null | undefined,
): string[] {
  const names = crewNamesForRole(list, role);
  const trimmed = (current ?? "").trim();
  if (trimmed && !names.some((n) => crewNamesEqual(n, trimmed))) {
    return [trimmed, ...names];
  }
  return names;
}

File: src-tauri/src/model/crew.rs
//! Crew roster for TM/VS prediction (Phase 19a).
//!
//! Mirrors ATS `CrewMember` / `DEFAULT_CREW_LIST`, plus editable `aliases`.

use serde::{Deserialize, Serialize};

/// One person in the local crew roster.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CrewMember {
    pub name: String,
    #[serde(default)]
    pub tandemmaster: bool,
    #[serde(default)]
    pub videospringer: bool,
    /// Alternate spellings / short forms matched in folder names (e.g. Corni → Cornelius).
    #[serde(default)]
    pub aliases: Vec<String>,
}

impl CrewMember {
    pub fn new(name: impl Into<String>, tandemmaster: bool, videospringer: bool) -> Self {
        Self {
            name: name.into(),
            tandemmaster,
            videospringer,
            aliases: Vec::new(),
        }
    }

    pub fn with_aliases(mut self, aliases: impl IntoIterator<Item = impl Into<String>>) -> Self {
        self.aliases = aliases.into_iter().map(Into::into).collect();
        self
    }

    /// Case-insensitive equality against canonical name or any alias.
    pub fn matches_token(&self, token: &str) -> bool {
        let t = token.trim();
        if t.is_empty() {
            return false;
        }
        if eq_ci(&self.name, t) {
            return true;
        }
        self.aliases.iter().any(|a| eq_ci(a, t))
    }
}

fn eq_ci(a: &str, b: &str) -> bool {
    a.trim().eq_ignore_ascii_case(b.trim())
}

/// Production roster from ATS `DEFAULT_CREW_LIST`, plus sensible start aliases.
pub fn default_crew_list() -> Vec<CrewMember> {
    let mut list = vec![
        CrewMember::new("Alberto", true, false),
        CrewMember::new("Ana", true, true),
        CrewMember::new("Andy", true, true).with_aliases(["Andreas"]),
        CrewMember::new("Chris", true, false),
        CrewMember::new("Cornelius", true, false).with_aliases(["Corni"]),
        CrewMember::new("Futti", true, true),
        CrewMember::new("Harry", true, true),
        CrewMember::new("Henrik", true, true).with_aliases(["Henni"]),
        CrewMember::new("Jan", true, false),
        CrewMember::new("Jojo", false, true),
        CrewMember::new("Kai", false, true),
        CrewMember::new("Käthe", false, true),
        CrewMember::new("Mathi", true, true).with_aliases(["Mathias"]),
        CrewMember::new("Max", true, false),
        CrewMember::new("Mayo", true, true),
        CrewMember::new("Pascal", true, false).with_aliases(["Passy"]),
        CrewMember::new("Ralph", true, true),
        CrewMember::new("Rene", true, false),
        CrewMember::new("Robert", false, true),
        CrewMember::new("Robin", false, true),
        CrewMember::new("Sabrina", false, true),
        CrewMember::new("Sahira", true, true),
        CrewMember::new("Samuel", true, true).with_aliases(["Samu"]),
        CrewMember::new("Stefan", true, false),
        CrewMember::new("Steve", true, false),
        CrewMember::new("Tim", true, true),
        CrewMember::new("Tom", true, true),
        CrewMember::new("Torsten", true, true),
    ];
    list.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));
    list
}

/// Parse `crew_list` setting JSON; empty / invalid → defaults.
pub fn load_crew_list(raw: &str) -> Vec<CrewMember> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return default_crew_list();
    }
    match serde_json::from_str::<Vec<CrewMember>>(trimmed) {
        Ok(list) if !list.is_empty() => list,
        _ => default_crew_list(),
    }
}

/// Serialize crew list for the `crew_list` setting key.
pub fn serialize_crew_list(list: &[CrewMember]) -> Result<String, serde_json::Error> {
    serde_json::to_string(list)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_crew_has_expected_roles_and_aliases() {
        let list = default_crew_list();
        assert!(list.iter().any(|c| c.name == "Andy" && c.tandemmaster && c.videospringer));
        assert!(list.iter().any(|c| c.name == "Stefan" && c.tandemmaster && !c.videospringer));
        assert!(list.iter().any(|c| c.name == "Robin" && !c.tandemmaster && c.videospringer));
        let cornelius = list.iter().find(|c| c.name == "Cornelius").expect("Cornelius");
        assert!(cornelius.tandemmaster);
        assert!(!cornelius.videospringer);
        assert!(cornelius.matches_token("Corni"));
        assert!(cornelius.matches_token("corni"));
        assert!(list.iter().find(|c| c.name == "Samuel").unwrap().matches_token("Samu"));
        assert!(list.iter().find(|c| c.name == "Pascal").unwrap().matches_token("Passy"));
        assert!(list.iter().find(|c| c.name == "Andy").unwrap().matches_token("Andreas"));
        assert!(list.iter().find(|c| c.name == "Henrik").unwrap().matches_token("Henni"));
        assert!(list.iter().find(|c| c.name == "Mathi").unwrap().matches_token("Mathias"));
    }

    #[test]
    fn load_crew_list_falls_back_on_empty_or_invalid() {
        assert_eq!(load_crew_list(""), default_crew_list());
        assert_eq!(load_crew_list("not-json"), default_crew_list());
        assert_eq!(load_crew_list("[]"), default_crew_list());
    }

    #[test]
    fn load_crew_list_roundtrips_custom() {
        let custom = vec![
            CrewMember::new("Ada", true, false).with_aliases(["A"]),
            CrewMember::new("Bea", false, true),
        ];
        let json = serialize_crew_list(&custom).unwrap();
        let loaded = load_crew_list(&json);
        assert_eq!(loaded, custom);
    }

    #[test]
    fn aliases_default_when_missing_in_json() {
        let raw = r#"[{"name":"Tom","tandemmaster":true,"videospringer":true}]"#;
        let loaded = load_crew_list(raw);
        assert_eq!(loaded.len(), 1);
        assert!(loaded[0].aliases.is_empty());
    }
}
